File: solutions/ruby/saddle-points/saddle_points.rb
module Grid
  def self.row_maxs(input)
    input.zip(input.map(&:max)).map do |row, max|
      row.map.with_index.select { |v, _| v == max }.map { |_, i| i }
    end
  end

  def self.col_mins(input)
    cols = input.transpose
    cols.zip(cols.map(&:min)).map do |col, min|
      col.map.with_index.select { |v, _| v == min }.map { |_, i| i }
    end
  end

  def self.saddle_points(input)
    row_max = row_maxs(input)
    col_min = col_mins(input)

    points = []
    input.each_index do |row|
      input[0].each_index do |col|
        if row_max[row].include?(col) && col_min[col].include?(row)
          points.append({ 'row' => row + 1, 'column' => col + 1 })
        end
      end
    end
    points
  end

  private_class_method :row_maxs, :col_mins

  public_class_method :saddle_points
end

File: solutions/python/saddle-points/saddle_points.py
def saddle_points(matrix):
    # Empty
    if not matrix:
        return []
    columns = len(matrix[0])
    # Not uniform shape
    if not all(len(r) == columns for r in matrix):
        raise ValueError("irregular matrix")

    col_minimums = [
        min(matrix[i][col] for i in range(len(matrix)))
        for col in range(columns)
    ]

    points = []
    for row_idx, row in enumerate(matrix):
        row_max = max(row)
        for col_idx, val in enumerate(row):
            if val == row_max and val == col_minimums[col_idx]:
                points.append({"row": row_idx + 1, "column": col_idx + 1})
    return points
